<div class="docbook-page">
  <div style="clear:" class="box">
    <div class="heading2">Tutorial 1: Creating and using Lists</div>
    <div class="body">
      <p style="">All queries in FlyMine can be performed on lists of objects
    that you create yourself. Saved lists of data can be used to
    constrain queries (either in a template query or in the
    QueryBuilder), perform logical operations (union, intersect,
    subtract) and can be exported.</p>
      <ul>
        <li>Lists can be created from an external source:
      click on 'Lists' on the top menu bar of any FlyMine page and
      then click on 'Upload'. This page allows you to either
      type/paste in a list of identifiers, or upload a file containing
      the list.<br/>
</li>
        <li>Data from previous queries can also be saved
      into a list. To save your data to a list select the required data
      on a result page and either add it to an existing list or save it
      to a new list, using the boxes displayed underneath the
      results. Selecting the check boxes next to the column headings
      will select the entire column for saving. Alternatively,
      individual items can be saved by selecting check boxes next to
      the items.<br/>
</li>
      </ul>
      <p style="">This tutorial will illustrate how to create lists and how
      to make use of lists. As an example, we will search for the
      orthologues of <i>D. melanogaster</i> eye
      specification genes in <i>D. pseudoobscura</i>. We
      will then export the gene sequences for all the orthologues at
      once. To find the orthologues of the
      <i>D. melanogaster</i> genes in
      <i>D. pseudoobscura</i>, we will first make a list
      of the <i>D. melanogaster</i> genes. Click on
      'Lists' in the top menu bar, and then on the list 'Upload' tab
      to create a list from an external source.</p>
      <img alt="[toolbar_upload.png]" class="inlinegraphic" src="toolbar_upload.png"/>
      <address> </address>
      <p style="">To make your list:</p>
      <ol>
        <li>Since it is currently not possible to create a
      list of mixed objects (eg proteins and genes), you have to
      specify your type of list. In this example, we will make a list
      of type 'Gene'. It is possible, however, to have a mix of the
      type of identifier (eg for a list of Drosophila genes your list
      may contain both gene symbols, CG identifiers and
      FBgns). <br/>
</li>
        <li>Optionally you can also set the organism for
      your list of identifiers - this is useful if you have, for
      example, gene symbols in your list which may appear in more than
      one organism. In this example, select <i>Drosophila
      melanogaster</i>. <br/>
</li>
        <li>Now you can either type or cut and paste in a
      list of identifiers, or upload a list from a file. Separate
      identifiers by a comma, space, tab or new line.  Qualify any
      identifiers that contain whitespace with double quotes like so:
      "even skipped". As an example, type in the following 10 gene
      symbols in the box: argos, PAX6_DROME, CG9554, fbgn0011647, lz,
      toy, dac, toe, "sine oculis", Dsix3 and click on 'Create
      list'. Before creating your list, all your identifiers will be
      compared against the FlyMine database and any discrepancies will
      be reported to you on the 'list confirmation
      page'<br/>
</li>
      </ol>
      <img alt="[list_upload.png]" class="inlinegraphic" src="list_upload.png"/>
      <address> </address>
      <p style="">List confirmation page:</p>
      <p style="">The example list check matches 7 genes successfully but contains 3
discrepancies - two low quality matches and a converted type:</p>
      <ol>
        <li>- Low quality matches: A low quality match is
      one that does not match the main identifiers in the database but
      which was found by searching the synonyms. In this case Dsix3
      and FBgn0011647 are old (or secondary) identifiers. However, it
      has found the genes that we want so we can add both to our list
      by clicking on 'add' in the last column.<br/>
</li>
        <li>- Converting types: When we uploaded our list we
      specified that we wanted to make a list of genes. However one
      item in our list is actually a protein identifier
      (PAX6_DROME). FlyMine has found this protein in the database
      along with the gene that encodes it. It is not possible to make
      a list of mixed 'types' but we can now add the gene for this
      protein to our list.<br/>
</li>
        <li>- Duplicate genes: You may also see
      discrepancies called 'Duplicate genes'. This can occur if, for
      example, you have gene symbols in your list and the symbol is
      found in more than one organism. Such discrepancies can be
      avoided by using the optional organism constraint on the list
      upload page. It will also happen when two gene symbols differ in
      case, since FlyMine is not case-sensitive, eg . In any of these
      case, all occurrences of the gene will be listed and you can
      select the one you intended to be in your list. Once you are
      happy with your list you can give it a name, eg. eye_genes_dmel,
      and press 'Save list'.<br/>
</li>
      </ol>
      <img alt="[list_confirm.png]" class="inlinegraphic" src="list_confirm.png"/>
      <address> </address>
      <p style="">Details of the contents of a list together with additional
     information about a list can be viewed through the 'List
     Analysis' page. The analysis page for a list is automatically
     displayed when a list is created. List analysis pages can be
     accessed at other times by clicking on the list name on the 'list
     page'.</p>
      <p style="">All list analysis pages have the same structure and consist of three sections:</p>
      <ol>
        <li>- a summary of the contents of the list<br/>

        
          
            <img alt="[listpage_top.png]" class="inlinegraphic" src="listpage_top.png"/>
          <address> </address>
          

</li>
        <li>- a series of widgets giving more information on
      the list, for example, for a list of genes, a graph of the
      chromosome distribution, a statistical analysis of the GO term
      distribution (GO term enrichment), KEGG pathway data, protein
      domain enrichment and publication enrichment.<br/>

        
          
            <img alt="[listpage_widget.png]" class="inlinegraphic" src="listpage_widget.png"/>
          <address> </address>
          
</li>
        <li>- a set of template queries that have been
      pre-run on the set of objects in the list<br/>
</li>
      </ol>
      <img alt="[listpage_templates.png]" class="inlinegraphic" src="listpage_templates.png"/>
      <address> </address>
      <p style="">Now we want to find all the orthologous genes in
    <i>D. pseudoobscura</i>. We can use an existing
    template for that. The template can be found in two different
    ways:</p>
      <ol>
        <li>Select the Comparative Genomics aspect from the
      FlyMine homepage and find the template 'Gene --&gt; Orthologues in
      one specific organism'.<br/>
</li>
        <li>Alternatively, go to the Templates page and find
      the right template using the search box.<br/>
</li>
      </ol>
      <img alt="[template_filter.png]" class="inlinegraphic" src="template_filter.png"/>
      <address> </address>
      <p style="">In both cases, the template can be opened by clicking on the template title.</p>
      <p style="">Any variable in a template query can be constrained
    to a list (or 'not in' a list) rather than a single value. Click
    on the check-box below the Gene box to constrain the template to
    all genes in a list. The name of the list you just created will
    appear in the drop-down menu. Select the right organisms from the
    two other drop-down menus and run the query.</p>
      <img alt="[list-template.png]" class="inlinegraphic" src="list-template.png"/>
      <address> </address>
      <p style="">The results table shows the
   <i>D. melanogaster</i> genes with their corresponding
   orthologues in <i>D. pseudoobscura</i>. Only 7 out of
   the 10 <i>D. melanogaster</i> genes have orthologues
   in <i>D. pseudoobscura</i>.</p>
      <p style="">Save the orthologous genes in a new list:</p>
      <ul>
        <li>Click on the check-box in the title of column 4
      which contains the gene symbols for the orthologues. This
      selects all items in that column (on every result page, if more
      than one page) plus in three additional columns which refer to
      the same object. This is because you are saving the actual item
      (eg the gene) to a list, not just the
      identifier.<br/>
</li>
        <li>At the bottom of the page is an option to save
      the selected results in a new list. Choose a new list name
      eg. 'eye_genes_dpse' and click 'Save'.<br/>
</li>
      </ul>
      <img alt="[list-template-results.png]" class="inlinegraphic" src="list-template-results.png"/>
      <address> </address>
      <p style="">Now we want to download the sequences of both the
    <i>D. melanogaster</i> genes and the
    <i>D. pseudoobscura</i> genes. The quickest way to
    achieve this is to fuse the two lists and download all sequences
    at once: </p>
      <ul>
        <li>Go to the Lists-View page. Select the two lists
      (eyes_genes_dmel and eyes_genes_dpse), type in a new name for
      the new combined list and click on 'union'. A message will
      appear saying that a new list was created, and the list itself
      will be listed on top. <br/>

        
          
            <img alt="[list-union.png]" class="inlinegraphic" src="list-union.png"/>
          <address> </address>
        
</li>
        <li>Clicking on the new list name will display the
      list analysis page for this list. <br/>
</li>
        <li>Click on 'View all 17 records' to display it as
      a result page. At the bottom of the result page is a link called
      'Export first visible column in FASTA format'. This will
      download the sequences of all the 17 genes in your
      list.<br/>
</li>
      </ul>
      <img alt="[fasta-export.png]" class="inlinegraphic" src="fasta-export.png"/>
      <address> </address>
      <p style="">Lists of data, as well as previously executed queries, can be
    retrieved from your MyMine page at the top menu bar on any
    FlyMine page (lists can also be directly accessed on the
    'Lists-view' page). Your lists and query history will remain available to
    you only for the duration of a session unless you create and use a
    login. If you log in (via button on top menu bar), your lists and
    query history can be stored permanently.</p>
      <img alt="[list-MyMine.png]" class="inlinegraphic" src="list-MyMine.png"/>
      <address> </address>
      <p style="">Find your lists back in 'Lists'.</p>
      <img alt="[list-history.png]" class="inlinegraphic" src="list-history.png"/>
      <address> </address>
    </div>
  </div>
</div>
